Make ticket descriptions easier to read

tonyandrewmeyer opened the pull request:
Convert ticket descriptions from WikiFormatting to mrkdwn so that they are easier to read in Slack.

tonyandrewmeyer committed:
Convert descriptions to mrkdwn

Add a module that can convert text in Trac’s “WikiFormatting” markup
into markdown and mrkdwn (Slack’s markdown variant).

Use this module to convert ticket descriptions, in query and describe
results.

There are several places where conversion still needs to be done, and a
few where conversion isn’t really possible.

Note that I have only done limited testing of the conversion process.
It would be easier to do it in bulk in Slack itself :wink:.

tonyandrewmeyer committed:
Ignore .pyc files

alexkiro commented:
@tonyandrewmeyer this doesn't work unfortunalley. It seems that TracXML is not returning any `\n` in the description for some reason. :( So the markdown is ignored because it only works with newlines and so on. 

I think we should try to get them from Trac somehow. Maybe use JSONRPC instead of XML. If not we should artificially insert them, but that wouldn't be good enough, for example the `{{{ }}}` parts.

alexkiro commented on trac_to_markdown.py:
````diff
@@ -0,0 +1,89 @@
+    # XXX This doesn't handle nested blocks, but those should be
+    # XXX uncommon. We would probably need to properly parse the text
+    # XXX to handle that.
+    text = re.sub(r"^{{{\s(.*?)\s}}}$", r"```\n\1\n```", text, re.DOTALL)
````
I think you needs a `re.M` here as well.

tonyandrewmeyer replied:
Ah, yes, thanks. I had that right in an earlier version (pre-committing) and then broke it when I re-wrote.

alexkiro commented on trac_to_markdown.py:
```diff
@@ -0,0 +1,89 @@
+    text = re.sub(r"(^\s\*)(\s)", transforms["ul"], text)
+    text = re.sub(r"(^\s\-)(\s)", transforms["ul"], text)
+    text = re.sub(r"^\s(\d+)\.(\s)", transforms["ol"], text)
+
```
Also these 3 seems to need `flags=re.M` since the `^` will only match at the start of string otherwise.

tonyandrewmeyer replied:
Ah yes, because we're not doing this line by line.  Thanks!

alexkiro commented on trac_to_markdown.py:
````diff
@@ -0,0 +1,89 @@
+    # XXX This doesn't handle nested blocks, but those should be
+    # XXX uncommon. We would probably need to properly parse the text
+    # XXX to handle that.
+    text = re.sub(r"^{{{\s(.*?)\s}}}$", r"```\n\1\n```", text, re.DOTALL)
````
Also I'm pretty sure that's the positional argument for `count` not `flags` in `re.sub` :) Fallen into this trap quite recently myself.

tonyandrewmeyer committed:
Fix flags argument.

count is the third argument, so use kwargs.

tonyandrewmeyer committed:
Fix lists to match multiple lines.

Lists will be at the start of multiple lines, so we need to have the
re.M flag.

tonyandrewmeyer committed:
Fix code comment replacement.

The code block may start at the start of any line, not just the first,
so also use the re.M flag.

tonyandrewmeyer commented:
> It seems that TracXML is not returning any \n in the description for some reason. :(

:( indeed. Will take a look at that then.

tonyandrewmeyer committed:
Add support for the JSON RPC

If jsonrpclib is available, then provide two additional classes (one
http and one https) that work with the JSON version of the RPC API.

tonyandrewmeyer committed:
Use the JSON version of the RPC API if possible.

If jsonrpclib is available, then use the JSON version of the RPC API
rather than the XML one.

The XML version seems to lose \n (e.g. in descriptions) for some
reason. It’s not clear where this bug comes from - most likely the
xmlrpc plug-in for Trac.

I have not tested this extensively, but general retrieval of ticket
details does work, so everything else ought to be fine in theory!

tonyandrewmeyer commented:
In theory all the things you identified are fixed. I have not tested this all together anywhere, though, just tested individual bits as I worked on them.

alexkiro commented:
So it doesn't get lost in Slack, there's still a problem with date parsing: 

```
2016-09-09 14:56:35,924 ERROR Exception on /trac-slack [POST]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python2.7/site-packages/flask/views.py", line 84, in view
    return self.dispatch_request(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/flask/views.py", line 149, in dispatch_request
    return meth(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/mimerender.py", line 244, in wrapper
    result = target(*args, **kwargs)
  File "/home/spamexperts/trac-slack/application.py", line 263, in post
    return self._handle_describe("id=%s" % query)
  File "/home/spamexperts/trac-slack/application.py", line 198, in _handle_describe
    attr = self._get_tick_attributes(ticket)
  File "/home/spamexperts/trac-slack/application.py", line 161, in _get_tick_attributes
    stamp = calendar.timegm(attributes['time'].timetuple())
AttributeError: 'dict' object has no attribute 'timetuple'
```

tonyandrewmeyer committed:
Handle dates in the response.

The Trac XMLRPC plug-in has a custom format for returning types that
JSON cannot handle (specifically datetime and binary).

Convert any datetime objects that are received to actual
datetime.datetime objects.

This could be expanded to binary objects as well, but we aren’t using
them at the moment, and I’m not sure what the destination type would be
(bytestring?) so leave that.

tonyandrewmeyer commented:
Although this probably works now, I'm wondering if a better approach might be to use `wiki.wikiToHtml(text)` and then a HTML to markdown conversion. It would end up as markdown rather than mrkdwn, but it might be nice enough to display (although links wouldn't be as good or would need more processing). A double-conversion seems messy, but most of the code would be third-party, which would be nice.

tonyandrewmeyer commented:
Actually, I guess the other thing I should have checked is the xmlrpc plug-in. We're using 1.1.2-r11439, and the latest is 1.1.5. 

There is mention of [this problem](https://trac-hacks.org/changeset/13776) although the fix is in code that we don't even have in our version.

So it's possible that we can go back to XML by just upgrading. I'll check that at some point and open a new PR if so.
